cadcdata/cadcdata/storageinv.py

```diff
@@ -0,0 +1,826 @@
+            """
+            Wrapper class to make response.raw.read work as iterator and behave
+            the same way as the corresponding response.iter_content
+            """
+
+            def __init__(self, rsp):
+                """
+                :param rsp: HTTP response object
+                """
+                self._read = rsp.raw.read
+                self._decode = rsp.raw._decode
+                self.block_size = 0
+                self._md5sum = net.extract_md5(rsp.headers)
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
+                return self.next()
+
+            def next(self):
+                # reads the next raw block
+                data = self._read(self.block_size)
+                if len(data) > 0:
+                    if self._md5sum:
+                        hash_md5.update(data)
+                    return data
+                else:
+                    raise StopIteration()
+
+            def get_instance(self, block_size):
+                self.block_size = block_size
+                return self
+
+        try:
+            total_length = int(response.headers.get('content-length', 0))
+        except ValueError:
+            pass
+
+        rr = RawRange(response)
+        reader = rr.get_instance
+        if logger.isEnabledFor(logging.INFO):
+            if total_length != 0:
+                chunks = progress.bar(reader(
+                    READ_BLOCK_SIZE),
+                    expected_size=((total_length / READ_BLOCK_SIZE) + 1))
+            else:
+                chunks = progress.mill(reader(READ_BLOCK_SIZE),
+                                       expected_size=0)
+        else:
+            chunks = reader(READ_BLOCK_SIZE)
+        start = time.time()
+        for chunk in chunks:
+            if process_bytes is not None:
+                process_bytes(chunk)
+            dest_file.write(chunk)
+        src_md5sum = net.extract_md5(response.headers)
+        if src_md5sum:
+            dest_md5sum = hash_md5.hexdigest()
+            if src_md5sum != dest_md5sum:
+                raise DownloadError(
+                    'Downloaded file is corrupted: '
+                    'expected md5({}) != actual md5({})'.
+                    format(src_md5sum, dest_md5sum))
+        duration = time.time() - start
+        logger.info(
+            'Successfully downloaded file {} as {} in {}s (avg. speed: {}MB/s)'
+            ''.format(resource, dest_file.name, round(duration, 2),
+                      round(total_length / 1024 / 1024 / duration, 2)))
+
+    def cadcput(self, id, src, replace=False, file_type=None,
+                file_encoding=None, md5_checksum=None):
+        """
+        Puts a file into the inventory system
+        :param id: unique identifier (URI) for the file in the CADC inventory
+        system
+        :param src: location of the source file
+        :param replace: boolean indicated whether this is expected to be
```

Comment:
I thought about it and I've added the flag in the last minute (it was part of `adPut` and `cadc-data get`). The purpose of the flag is to safeguard the data (sometimes the unique copy) from accidental deletes. Because a `PUT` to a wrong URI could accidentally wipe out existing data without a possibility of recovering it. And a wrong URI could be just the result of a typo for example.
Also, the bulk of our requests are `GET`s. The model is that we `PUT` once (or twice) into the archive but serve multiple times so the time saved by not calling `INFO` first is negligible. If you have and other users have a strong argument for this we could implement the replace with options `[yes|no|maybe]` , with `maybe` being doesn't matter (because the definitive copy is elsewhere).

Reply:
I think if you add the reasoning in your reply to the original comment, to the code comments, that will be sufficient.

Reply:
Added reasoning.